### DISCUSSION TITLE
RP2040-Zero + Level shifter

### DISCUSSION THREAD

**Discussion by AstaRom** [awawa-dev/HyperSPI#54] [Show and tell]
Good day @awawa-dev 

I decided to try the rp2040 board in my assembly (the main board at the moment is esp32).
I sketched out the circuit, soldered everything, and launched it for testing.

With the following connection diagram everything worked fine:

- the board 'rp2040' was powered via Type-C;
- the 'level shifter' received power from the 'LED strip';
- Common ground line between 'SN74AHCT125N', 'LED strip', 'rp2040' board and 'RPi4'.

The rp2040 specification states that the board can be powered directly through pins on the board (5V, GND), rather than just via USB.
I decided to try it out and powered the board directly from the LED strip unit, similar to a level shifter.

- the board 'rp2040' received power from the 'LED strip';
- the 'level shifter' received power from the 'LED strip';
- Common ground line between 'SN74AHCT125N', 'LED strip', 'rp2040' board and 'RPi4'.

<details>
  <summary>Connection diagram</summary></br>

![main_circuit](https://github.com/user-attachments/assets/99420d8b-d521-475c-97e8-4031326d4b9a)</br>
![circuit_board](https://github.com/user-attachments/assets/4afdcf52-06dc-447b-8c70-669c8a0cb5ab)</br>
![preview](https://github.com/user-attachments/assets/616a7694-055b-418f-a85e-4835d602a8d9)

</details>

With this scheme, everything stops working, 3 to 4 diodes in the strip periodically blink.
The 'rp2040' board does not respond in any way to commands from the HyperHDR interface.

Have you by any chance tried this type of connection?
If so, did everything work fine?

**Comment by awawa-dev** [comment]
Hi
Sorry, I've no clue :( I don't use power by pinout solution because I dont trust such connections but it's my experience from Raspberry Pi SBC boards not Pico. As far I know 5v pin on Zero board is VSYS so you can power the board using external power supply when the USB-C is disconnected (https://www.penguintutor.com/electronics/pico-power). Unless they connected it to VBUS directly during the cost "optimization" process and then the LED power supply may be out of the tolerance.

> 3 to 4 diodes in the strip periodically blink.

if these LEDs blink even when HyperHDR is not working that's the mayor problem. Something triggers data line and it's not HyperSPI because it's using data integrity check. Maybe it's a common ground problem somehow (loops)?

**Reply by awawa-dev** [reply]
Check if the Pico voltage regular is working: measure 3v3 pin?

**Reply by AstaRom** [reply]
> Unless they connected it to VBUS directly during the cost "optimization" process

Maybe.

> Maybe it's a common ground problem somehow (loops)?

The connections on the PCB are the same in both cases.
In the first one, I just need to connect the Type-C cable and not connect one +5V pin from the power supply. Everything works.
We turn off Type-C and connect the +5V pin - it doesn’t work.

> Check if the Pico voltage regular is working: measure 3v3 pin?

Tomorrow evening I can check, I don’t have a multimeter at hand at the moment :)

**Reply by awawa-dev** [reply]
> > Maybe it's a common ground problem somehow (loops)?
> 
> The connections on the PCB are the same in both cases. In the first one, I just need to connect the Type-C cable and not connect one +5V pin from the power supply. Everything works. We turn off Type-C and connect the +5V pin - it doesn’t work.
> 

there is one difference: the USB connector also serves as an additional grounding connector between Rpi and Pico, which is missing when you switch to pin-based power. Although I don't know if that matters, we're in somewhat uncharted territory. this LED blinking looks to me a bit like parasitic power supply by data channel via internal diode (resistor can eliminate this) but I have no idea why it only happens when you connect power via pinout...

**Reply by AstaRom** [reply]
> there is one difference: the USB connector also serves as an additional grounding connector between Rpi and Pico

I connect USB from a separate source (power bank), not to RPi.
That's the point :)

> which is missing when you switch to pin-based power

The ground connection between the elements directly on the board, I rang with a multimeter earlier.
I observed the blinking of one diode in the absence of GND connection on esp32, where the entire tape worked normally except for this one diode.

Maybe the problem is in the board itself, in its operation when connecting power directly to the contacts (5V, GND).
I don’t know how to check the board itself and how it works.
I was only able to install the driver on it and run the effect in the HyperHDR window.

On esp32 I could install WLED for testing and check its operation, but this cannot be done here.

If I were more advanced, it would be possible to write some kind of verification code on the board that would complete some simple action (tape flickering) from output 14/15, then it would be possible to check the “rp2040” itself.

**Reply by awawa-dev** [reply]
you can add following simple debug code to the critical section you want to debug
```
     gpio_init(SOME_PIN);
     gpio_set_dir(SOME_PIN, GPIO_OUT);
..........
..........
     gpio_put(SOME_PIN, 1);
     sleep_ms(delay);
     gpio_put(SOME_PIN, 0);
     sleep_ms(delay);
```
but still you need to have a multimeter ;)

**Reply by AstaRom** [reply]
I tried the demo code to check.
- [fill_demo.ino](https://github.com/MrYsLab/NeoPixelConnect/blob/master/examples/fill_demo/fill_demo.ino)

I indicated pin 16 in the code (LED on the board).
I connected the board with power through the contacts (5V, GND), nothing works.
Once you connect the USB, the tape flash in three colors (according to the example).

Probably the problem is in the board itself.
Maybe I got some kind of defective copy when I bought it. :))

**Reply by awawa-dev** [reply]
Hmm...```The VSYS pin of the RP2040 is connected to the VUSB pin directly in RP2040-zero (named Pin23 )``` (https://www.waveshare.com/wiki/RP2040-Zero)
This is a significant difference compared to the standard Pico board. So the input must be very stable providing 5V.
Usually
VSYS: 1.8 to 5.5V
but since it's connected to VBUS directly
VBUS: 5V ±10%

**Reply by AstaRom** [reply]
I'll order another board. I was wondering if I had received a defective board (It doesn't cost much, $1.8). 
So I liked it, it’s very compact and in the form that I soldered together with the level shifter it is 2-2.5 times smaller than Esp32.

<details>
  <summary></summary></br>

![rp2040-zero](https://github.com/user-attachments/assets/7379912a-efb7-4008-953a-5e482dda1689)

</details>

Later I will measure the pin 3.3V output with multimeters.

**Reply by awawa-dev** [reply]
If you have a spare usb2.0 cable you can do an experiment and after removing the cable plug and accessing the individual wires connect the 5v cable to the 5v pin and the ground from that cable to the board. Then provide a power supply (ex. from the power bank). If this works, the problem lies in the unstable power supply. Maybe a decoupling capacitor will help a bit? Unfortunately, I'm not an electronics engineer ;)

**Reply by AstaRom** [reply]
Before I forget.
The translation files are missing:

- "edt_prevent_pixel_data_splitting_across_universes".

<details>
  <summary></summary></br>

![Screenshot_1](https://github.com/user-attachments/assets/21b48607-0a4c-4dcd-826e-be1883ffb205)

</details>
I wanted to translate but I don’t know what this point means.

**Reply by awawa-dev** [reply]
edt_prevent_pixel_data_splitting_across_universes?
As far as remember it is related to this issue: https://github.com/awawa-dev/HyperHDR/issues/657
Probably it was lost while merging my changes.
maybe "Block color splitting between universes"?

**Reply by AstaRom** [reply]
Request is ready

**Reply by awawa-dev** [reply]
Thank you :)

**Reply by AstaRom** [reply]
> If you have a spare usb2.0 cable you can do an experiment and after removing the cable plug and accessing the individual wires connect the 5v cable to the 5v pin and the ground from that cable to the board. Then provide a power supply (ex. from the power bank). If this works, the problem lies in the unstable power supply. Maybe a decoupling capacitor will help a bit? Unfortunately, I'm not an electronics engineer ;)

Today I received a new board. I installed it and everything worked fine.
The RP2040 receives power through the contacts (5V, GND) on the board and HyperHDR can now control it normally.

It turns out that I received the first board with a manufacturing defect/malfunction.

**Reply by awawa-dev** [reply]
Did you receive exactly the same model?

**Reply by AstaRom** [reply]
> Did you receive exactly the same model?

Yes, exactly the same board, in appearance. The only difference is that I bought it from a different seller.

**Reply by awawa-dev** [reply]
Strange. Although I've heard of problems with some Pico Zeros powered directly but by 3.3V (installed for nefarious purposes in certain consoles). So maybe their quality matters in certain scenarios.

**Reply by AstaRom** [reply]
Perhaps, in the first board purchased, some element (resistor, capacitor, etc.) was not soldered on it or was poorly soldered. With conveyor assembly, no one is immune from this. I quickly examined both, it seems that everything that is on the first one is also on the second one, but whether they are welded well or not, most likely, can only be seen through a microscope.

**Comment by AstaRom** [comment]
While I had a free minute, I made a sketch of the printed circuit board. 
<details>
  <summary>Example</summary></br>

![Screenshot_1](https://github.com/user-attachments/assets/1387d327-6b24-4cef-8294-03def8be60a4)

</details>

It's easier to put together than the example in the title of the discussion, but it's less compact.

I think these examples may be useful to users of the HyperHDR community.

**Reply by TechsavvyCNC** [reply]
So you only have to connect 5 volts to the one pin on the SN74AHCT125N?

**Reply by AstaRom** [reply]
Top block, top to bottom: +5, Led data 1, Led data 2, -5 (Gnd).

From these pins, power is supplied via jumpers on the board to the level shifter and to rp2040.

**Reply by TechsavvyCNC** [reply]
So I should connect the pi to the pi zero like this image.  Also wire the SN74AHCT125N like the picture?

**Reply by AstaRom** [reply]
![rp2040](https://github.com/user-attachments/assets/36655dcf-f8c0-4f86-9f34-1d6c26bb1345)

**Comment by TechsavvyCNC** [comment]
[
![single_segment~2](https://github.com/user-attachments/assets/1f458c19-b33c-4943-b3aa-a73aac56f1fa)
]

**Comment by TechsavvyCNC** [comment]
Is esp32 or rp2040 better?

**Reply by AstaRom** [reply]
Both options have the right to life. Esp32 can be used with WLED; rp2040 can be selected only for a wired connection, the board is more compact.

**Comment by TechsavvyCNC** [comment]
So I should wire my esp32 like this picture?
![single_segment~2](https://github.com/user-attachments/assets/fe997db8-bc8a-4ea1-8385-ca9205507e86)

I don't need to plug it into usb and I shoud use firmware hypserspi right?

**Reply by AstaRom** [reply]
Esp32 needs to be powered from USB.
The above example is for the RP2040-Zero board.

**Reply by TechsavvyCNC** [reply]
The image above of the esp32 is right?  Use io 18,19, 23 and 5.  Also use USB to power it.

**Reply by AstaRom** [reply]
Yes, the connection diagram above is correct

**Reply by AstaRom** [reply]
19 pin is not necessary to connect now.

**Comment by AstaRom** [comment]
@awawa-dev , good evening.
I recently decided to try out the 'SN74LVC2T45' level shifter in a more compact package.
It seemed to me more convenient to work with it, I didn’t have to build extra jumpers on the board.
But it’s probably not suitable for everyone, it’s very small in size and it’s not so easy to solder.
- https://github.com/AstaRom/AstaRom.github.io/discussions/3
<details>
  <summary>Connection diagram</summary></br>
<img src="https://raw.githubusercontent.com/AstaRom/AstaRom.github.io/refs/heads/main/resources/img/pages/level_shifter_1.jpg" width=340px/>
</details>

Unlike you, I cannot see the signal at the output (for this, as I understand it, you need an oscilloscope), is there a big difference compared to the 'SN74AHCT125'.
